### ISSUE TITLE
Add an additional channel to the output layer

### ISSUE THREAD

**Issue by naush01** [ultralytics/yolov5#11817]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I would like to do object detection and depth estimation at the same time by adding an extra channel to the output layer. (Reference: https://www.mdpi.com/2079-9292/11/1/76)

Hence, there is an additional column in both of my training data and predicted value. However, I have no idea how to modify the architecture.

I have traced train.py on coco128 and printed out the items of the train_loader
`   
        pbar = enumerate(train_loader)

        LOGGER.info(('\n' + '%11s' * 7) % ('Epoch', 'GPU_mem', 'box_loss', 'obj_loss', 'cls_loss', 'Instances', 'Size'))
        if RANK in {-1, 0}:
            pbar = tqdm(pbar, total=nb, bar_format=TQDM_BAR_FORMAT)  # progress bar
        optimizer.zero_grad()

        for i, (imgs, targets, paths,_) in pbar:  # batch -------------------------------------------------------------
            # print("===================")
            # print("i=", i)
            # print("imgs", imgs)
            # print("paths = ", paths)
            # print("targets = ", targets)
            # print("===================")`

However, they can not match with the labels folder in coco128.

Consequently, I don't know how to deal with an additional column in my training dataset...

Thank you!

**Comment by glenn-jocher** [comment]
@naush01 hi there,

Thank you for reaching out to us! It seems like you want to add an additional channel to the output layer in order to perform both object detection and depth estimation simultaneously. However, modifying the YOLOv5 architecture to accommodate this extra channel can be a challenging task.

The YOLOv5 architecture is specifically designed for object detection, and making modifications to it requires a good understanding of the underlying code. To add an additional channel, you would need to modify the model architecture and potentially the loss function as well.

We suggest referring to the YOLOv5 model architecture in the `models` directory, specifically the `yolo.py` file. Here, you can modify the architecture to include the extra channel. However, keep in mind that this is a complex task and may require advanced knowledge of deep learning frameworks like PyTorch.

Additionally, you would also need to modify the label format to include the additional column. You mentioned that the label folder in `coco128` does not match with the items printed from the `train_loader`. It's important to ensure that your dataset labels are properly formatted to match the modified architecture.

We hope this information helps you get started with your modifications. However, please note that making substantial changes to the YOLOv5 architecture falls beyond the scope of our support. If you encounter any difficulties during the process, we recommend seeking assistance from the YOLO community or the Ultralytics team.

Good luck with your project, and let us know if you have any further questions!

**Comment by naush01** [comment]
@glenn-jocher Thanks for your quick reply!

Regarding the issue of the label folder in ```coco128``` does not match with the items printed from the ```train_loader```. Here is what I've done.

I followed the tutorial to download the ```coco128``` dataset (https://ultralytics.com/assets/coco128.zip) and put it under the ```dataset``` folder, and this ```dataset``` folder is next to the yolov5 directory.

Then, I printed out the items from the ```train_loader``` in ```train.py```
```
mloss = torch.zeros(3, device=device)  # mean losses
if RANK != -1:
    train_loader.sampler.set_epoch(epoch)
pbar = enumerate(train_loader)
LOGGER.info(('\n' + '%11s' * 7) % ('Epoch', 'GPU_mem', 'box_loss', 'obj_loss', 'cls_loss', 'Instances', 'Size'))
if RANK in {-1, 0}:
    pbar = tqdm(pbar, total=nb, bar_format=TQDM_BAR_FORMAT)  # progress bar
optimizer.zero_grad()

for i, (imgs, targets, paths,_) in pbar:  # batch -------------------------------------------------------------
    # print("===================")
    # print("i=", i)
    # print("imgs", imgs)
    # print("paths = ", paths)
    # print("targets = ", targets)
    # print("===================")
```

Afterward, I run the command ```python train.py --epochs 1 --batch-size 1```

Consequently, I checked the ```coco128``` folder by following the ```path```, and found that the ```target``` and ```label/img.txt``` are inconsistent. As you can see in the figure below, they are totally different.
<img width="719" alt="截圖 2023-07-06 上午9 46 26" src="https://github.com/ultralytics/yolov5/assets/118497994/30fbfc67-539b-4e61-9ad6-e95002dc95cf">

Moreover, I wonder how can we know the return value of ```train_loader``` is ```imgs, targets, paths,_```. I' ve traced the code in ```utils.dataloader.py``` but I could not find out that information.

Thank you for your patience.

**Comment by glenn-jocher** [comment]
@naush01 thank you for providing the detailed information. It seems that there might be a misunderstanding regarding the label and target structure in the `coco128` dataset.

The labels for the `coco128` dataset are present in the `labels` directory, while the images are in the `images` directory. The label files should have the same names as the respective image files but with a `.txt` extension. Each label file contains the object annotations in YOLO format, where each line represents one object in the image.

Regarding `train_loader`, the data returned is indeed in the format of `imgs, targets, paths, _` where:
- `imgs` represents the images
- `targets` contains the labels for the images
- `paths` is the path to the images
- `_` is a placeholder

It seems like there might be an issue with the dataset setup or data loading process. I would recommend double-checking the label and image file structure to ensure that they are properly aligned. Additionally, reviewing the `utils.datasets.py` file can provide insight into the return values of the `train_loader`.

If the issue persists, you may want to verify the dataset structure and loading logic to ensure that the labels and images are being correctly associated.

Let us know if you need further assistance. We're here to help!
